pkg/virt-operator/resource/apply/apps.go

```diff
@@ -87,7 +109,148 @@ func (r *Reconciler) syncDeployment(origDeployment *appsv1.Deployment) (*appsv1.
+	return getMaxUnavailable(daemonSet) == daemonSetDefaultMaxUnavailable.IntValue()
+}
+
+func (r *Reconciler) processCanaryUpgrade(cachedDaemonSet, newDS *appsv1.DaemonSet, forceUpdate bool) (bool, error, CanaryUpgradeStatus) {
+	var updatedAndReadyPods int32
+	var status CanaryUpgradeStatus
+	done := false
+
+	isDaemonSetUpdated := util.DaemonSetIsUpToDate(r.kv, cachedDaemonSet) && !forceUpdate
+	desiredReadyPods := cachedDaemonSet.Status.DesiredNumberScheduled
+	if isDaemonSetUpdated {
+		updatedAndReadyPods = r.howManyUpdatedAndReadyPods(cachedDaemonSet)
+	}
+
+	switch {
+	case updatedAndReadyPods == 0:
+		if !isDaemonSetUpdated {
+			// start canary upgrade
+			setMaxUnavailable(newDS, daemonSetDefaultMaxUnavailable)
+			_, err := r.patchDaemonSet(cachedDaemonSet, newDS)
+			if err != nil {
+				return false, fmt.Errorf("unable to start canary upgrade for daemonset %+v: %v", newDS, err), CanaryUpgradeStatusFailed
+			}
+		} else {
+			// check for a crashed canary pod
+			canary := r.getCanaryPod(cachedDaemonSet)
+			if canary != nil && util.PodIsCrashLooping(canary) {
+				r.recorder.Eventf(cachedDaemonSet, corev1.EventTypeWarning, failedUpdateDaemonSetReason, "daemonSet %v rollout failed", cachedDaemonSet.Name)
+				return false, fmt.Errorf("daemonSet %s rollout failed", cachedDaemonSet.Name), CanaryUpgradeStatusFailed
+			}
+		}
+		done, status = false, CanaryUpgradeStatusStarted
+	case updatedAndReadyPods > 0 && updatedAndReadyPods < desiredReadyPods:
+		if daemonHasDefaultRolloutStrategy(cachedDaemonSet) {
+			// canary was ok, start real rollout
+			setMaxUnavailable(newDS, daemonSetFastMaxUnavailable)
+			// start rollout again
+			_, err := r.patchDaemonSet(cachedDaemonSet, newDS)
+			if err != nil {
+				return false, fmt.Errorf("unable to update daemonset %+v: %v", newDS, err), CanaryUpgradeStatusFailed
+			}
+			log.Log.V(2).Infof("daemonSet %v updated", newDS.GetName())
+			status = CanaryUpgradeStatusUpgradingDaemonSet
+		} else {
+			log.Log.V(4).Infof("waiting for all pods of daemonSet %v to be ready", newDS.GetName())
+			status = CanaryUpgradeStatusWaitingDaemonSetRollout
+		}
+		done = false
+	case updatedAndReadyPods > 0 && updatedAndReadyPods == desiredReadyPods:
+		// rollout has completed and all virt-handlers are ready
+		// revert maxUnavailable to default value
+		setMaxUnavailable(newDS, daemonSetDefaultMaxUnavailable)
+		newDS, err := r.patchDaemonSet(cachedDaemonSet, newDS)
+		if err != nil {
+			return false, err, CanaryUpgradeStatusFailed
+		}
+		SetGeneration(&r.kv.Status.Generations, newDS)
+		log.Log.V(2).Infof("daemonSet %v is ready", newDS.GetName())
+		done, status = true, CanaryUpgradeStatusSuccessful
+	}
+	return done, nil, status
+}
+
+func getMaxUnavailable(daemonSet *appsv1.DaemonSet) int {
+	update := daemonSet.Spec.UpdateStrategy.RollingUpdate
+
+	if update == nil {
+		return 0
+	}
+	if update.MaxUnavailable != nil {
+		return update.MaxUnavailable.IntValue()
+	}
+	return 1
```

Comment:
Wouldn't `daemonSetDefaultMaxUnavailable` be more precise?

Reply:
Fixed.